src/wlserver.cpp

```diff
@@ -1083,6 +1084,208 @@ static void create_gamescope_control( void )
 	wl_global_create( wlserver.display, &gamescope_control_interface, version, NULL, gamescope_control_bind );
 }
 
+////////////////////////
+// commit-timing
+////////////////////////
+
+enum timer_stage {
+	TIMER_STAGE_LATCH = 0,
+	TIMER_STAGE_PRESENT = 1
+};
+
+enum timer_rounding_mode {
+	TIMER_ROUNDING_NEAREST = 0,
+	TIMER_ROUNDING_NOT_BEFORE = 1
+};
+
+struct commit_timer {
+	struct wl_resource *resource;
+
+	struct wlr_surface *surface;
+	enum timer_stage curr_stage;
+	enum timer_rounding_mode curr_rounding_mode;
+	uint64_t curr_timestamp_nsec;
+
+	struct wl_listener client_commit;
+	struct wlr_addon surface_addon;
+};
+
+static inline int64_t timespec_to_nsec(const struct timespec *a) {
+	return (int64_t)a->tv_sec * 1'000'000'000ul + a->tv_nsec;
+}
+
+static void resource_handle_destroy(struct wl_client *client, struct wl_resource *resource) {
+       wl_resource_destroy(resource);
+}
+
+static void commit_timer_destroy(struct commit_timer *timer) {
+	if (timer == NULL) {
+		return;
+	}
+	wl_list_remove(&timer->client_commit.link);
+	wlr_addon_finish(&timer->surface_addon);
+	wl_resource_set_user_data(timer->resource, NULL); // make inert
+	free(timer);
+}
+
+static void surface_timer_addon_handle_destroy(struct wlr_addon *addon) {
+	struct commit_timer *timer = wl_container_of(addon, timer, surface_addon);
+	commit_timer_destroy(timer);
+}
+
+static const struct wlr_addon_interface surface_timer_addon_impl = {
+	.name = "wp_commit_timer_v1",
+	.destroy = surface_timer_addon_handle_destroy,
+};
+
+#define NANOS_IN_SEC 1000000000
+static uint64_t timer_get_target_present_nsec(uint64_t timestamp_nsec, uint32_t rounding_mode) {
+	uint64_t refresh_nsec = NANOS_IN_SEC/(GetVBlankTimer().GetRefresh()/1000);
+	uint64_t cycle_phase_nsec = GetVBlankTimer().GetLastVBlank() % refresh_nsec;
+
+	uint64_t target_nsec = timestamp_nsec;
+	target_nsec -= cycle_phase_nsec;
+	if (rounding_mode == TIMER_ROUNDING_NEAREST) {
+		target_nsec += refresh_nsec/2;
+		target_nsec -= (target_nsec % refresh_nsec);
+	}
+	else if (rounding_mode == TIMER_ROUNDING_NOT_BEFORE) {
+		target_nsec = ((target_nsec+(refresh_nsec-1))/refresh_nsec)*refresh_nsec;
```

Comment:
I don't get why we are touching anything for `TIMER_ROUNDING_NOT_BEFORE`.

There is no world where we need to normalize it to a refresh cycle, and we probably shouldn't because multiple displays.

Reply:
Moved this logic to steamcompmgr, leaving out this case.